### PR TITLE
Move to in-docker CI builds and add Jenkins support

### PR DESCRIPTION
**Pull request by codablock**
This PR moves out all the real build logic from .travis.yml and puts it into their own scripts. These scripts are then called from Travis. The PR also adds 2 Jenkinsfiles, one for regular CI and one for Gitian building. The real work is done inside a Docker builder image, which is also built as part of the CI build.

The PR also implicitly upgrades us to GCC7 as the builder image is based on ubuntu:bionic. I had to backport a single commit from Bitcoin to fix back-compatibility with older systems.

I configured Jenkins to build all branches and PRs. However, only branches in this repo (no forks) are allowed to modify the Jenkinsfile itself. This is basically the main security measure against malicious PRs. This also means, that the Jenkinsfile should never execute any scripts directly without wrapping those into non-privileged docker containers (the builder image). The 2 Jenkins Jobs are configured to be only configurable by Jenkins users but viewable by everyone, including download of artifacts (e.g. Gitian binaries). Links to the publicly visible jobs:
https://jenkins.dash.org/blue/organizations/jenkins/dashpay-dash/activity
https://jenkins.dash.org/blue/organizations/jenkins/dashpay-dash-gitian-nightly/activity/

Moving to Jenkins required to change the way caching is handled in both Jenkins and Travis. Both will now put everything cache related into a single cache folder. For Travis, we continue using the native Travis caching features. For Jenkins this is more complicated and requires manual handling/copying of zipped artifacts.

Another notable change is that I moved all build-target related variables out of .travis.yml and moved these into matrix.sh, which is then sourced (instead of executed) whenever needed.

The push to Docker Hub is currently missing, but as Travis continues to run for now that should be fine. I'll add this later when we see that Jenkins is running stable.

### CODE REVIEW

**Comment by codablock** [comment]
There is a problem with the Travis build of the `linux64_nowallet` target. It keeps timing out because building Qt is horribly slow. I would assume that the same should happen on `develop` as well and already now, but caches are probably too good on develop since a very long time. So, if we merge this PR into develop, Travis will also start to timeout this build, as caches will be useless for the first build. To fix this, we might need to temporarily merge something into develop that causes the linux64_nowallet build to cancel earlier so that caches are filled up (which doesn't happen on timeout). And then remove the temporary change and hope the cache come to rescue. Or we disable that build for Travis forever and just count on Jenkins handling this from now on.

**Comment by codablock** [comment]
Another thing: This should be merged when develop gets open for 12.4 development.

File: ci/matrix.sh [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+export DOCKER_RUN_VOLUME_ARGS="-v $HOST_SRC_DIR:$SRC_DIR -v $HOST_CACHE_DIR:$CACHE_DIR"
+export DOCKER_RUN_ARGS="$DOCKER_RUN_VOLUME_ARGS -e PULL_REQUEST=$PULL_REQUEST -e JOB_NUMBER=$JOB_NUMBER -e BUILD_TARGET=$BUILD_TARGET"
+export DOCKER_RUN_IN_BUILDER="docker run -t --rm -w $SRC_DIR $DOCKER_RUN_ARGS $BUILDER_IMAGE_NAME"
+
+# Default values for targets
+export GOAL="install"
+export MAKEJOBS=-j5
+export SDK_URL=${SDK_URL:-https://bitcoincore.org/depends-sources/sdks}
+export PYTHON_DEBUG=1
+export MAKEJOBS="-j4"
```
Comment (UdjinM6) [inline ci/matrix.sh]: this vs line 28 ("-j4" vs j5)?

Comment (codablock) [reply ci/matrix.sh]: :see_no_evil: fixed

**Pull request merged**